torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp

```diff
@@ -745,6 +870,61 @@ void propagateQuantizationOps(Block* block) {
   }
 }
 
+void InsertQuantDeQuantHelper::runWeightObserver(
+    Module& module,
+    const std::string& method_name) {
+  for (auto& invoked_methods : getInvokedMethods(module, method_name)) {
+    auto& invoked_module = std::get<0>(invoked_methods);
+    const auto& invoked_method_name = std::get<1>(invoked_methods);
+    runWeightObserver(invoked_module, invoked_method_name);
+  }
+  Method method = module.get_method(method_name);
+  auto graph = method.graph();
+  Value* self = graph->inputs()[0];
+
+  std::vector<Value*> weight_values;
+  for (size_t idx = 1; idx < method.num_inputs(); ++idx) {
+    auto& v = graph->inputs()[idx];
+    if (v->type()->isSubtypeOf(TensorType::get())) {
+      auto observer_name = findObserverName(v);
+      // Ensure that we only track values that are the output of observer node.
+      // isWeight returns true if the observed value is a weight of quantized
+      // op.
+      if (observer_name && isWeight(module, v)) {
+        weight_values.push_back(v);
+      }
+    }
+  }
+  // Visit all blocks in the current graph to find weight values.
+  std::stack<Block*> blocks_to_visit;
+  blocks_to_visit.push(graph->block());
+  while (!blocks_to_visit.empty()) {
+    Block* b = blocks_to_visit.top();
+    blocks_to_visit.pop();
+    for (auto it = b->nodes().begin(), end = b->nodes().end(); it != end;) {
+      Node* n = *it++;
+      for (Value* v : n->outputs()) {
+        if (!v->type()->isSubtypeOf(TensorType::get())) {
+          continue;
+        }
+        auto observer_name = findObserverName(v);
+        if (observer_name && isWeight(module, v)) {
+          weight_values.push_back(v);
+        }
+      }
+      for (Block* subblock : n->blocks()) {
+        blocks_to_visit.push(subblock);
+      }
```

Comment:
can you check if this works for weight observers used in if?
e.g.
```python
def forward(self, x):
       if cond:
           x = F.linear(x, ...)
       else:
           x = F.linear(x, ...)
       return x
```

Reply:
It works for nn.Linear. I am looking into making it work for F.linear as well.

Reply:
the weight for nn.Linear is in the method graph, and there is no `if` in that graph.

Reply:
I think the main issue with `if` in the graph is caching the graph in `graph_to_subgraph_nodes_`. We call the extract code once for each weight value with the same graph input. Since this only provides a small optimization benefit I'm planning to remove this and make sure we correctly identify the weight subgraph even if the graph has multiple if statements. @jerryzh168 let me know if you think otherwise.

Reply:
I don't feel we need to remove the caching to support prim::If, why is that a problem?